lms/djangoapps/bookmarks/views.py

```diff
@@ -0,0 +1,278 @@
+
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import CourseKey, UsageKey
+
+from bookmarks.serializers import BookmarkSerializer
+from openedx.core.lib.api.serializers import PaginationSerializer
+
+from xmodule.modulestore.exceptions import ItemNotFoundError
+from xmodule.modulestore.django import modulestore
+
+from .models import Bookmark
+from .api import get_bookmark
+
+
+log = logging.getLogger(__name__)
+
+DEFAULT_FIELDS = ["id", "course_id", "usage_id", "created"]
+OPTIONAL_FIELDS = ['display_name', 'path']
+
+
+class BookmarksView(ListCreateAPIView):
+    """
+    **Use Case**
+
+        Get a paginated list of bookmarks in particular course.
+        Each page in the list can contain up to 30 bookmarks by default.
+
+        Create/Post a new bookmark for particular Xblock.
+
+    **Example Requests**
+
+          GET /api/bookmarks/v0/bookmarks/?course_id={course_id1}
+
+          POST /api/bookmarks/v0/bookmarks/?course_id={course_id1}
+
+    **Response Values**
+
+        * count: The number of bookmarks in a course.
+
+        * next: The URI to the next page of bookmarks.
+
+        * previous: The URI to the previous page of bookmarks.
+
+        * num_pages: The number of pages listing bookmarks.
+
+        * results:  A list of bookmarks returned. Each collection in the list
+          contains these fields.
+
+            * id: String. The identifier string for the bookmark: {user_id},{usage_id}.
+
+            * course_id: String. The identifier string of the bookmark's course.
+
+            * usage_id: String. The identifier string of the bookmark's XBlock.
+
+            * display_name: (optional) String. Display name of the XBlock.
+
+            * path: (optional) List. List of dicts containing {"usage_id": "", display_name:""} for the XBlocks
+                from the top of the course tree till the parent of the bookmarked XBlock.
+
+            * created: ISO 8601 String. The timestamp of bookmark's creation.
+
+    """
+    authentication_classes = (OAuth2Authentication, SessionAuthentication)
+    permission_classes = (permissions.IsAuthenticated,)
+
+    paginate_by = 30
+    paginate_by_param = 'page_size'
+    pagination_serializer_class = PaginationSerializer
+    serializer_class = BookmarkSerializer
+
+    def get_serializer_context(self):
+        """
+        Extra context provided to the serializer class.
+        """
+        context = super(BookmarksView, self).get_serializer_context()
+        if self.request.method == 'POST':
+            return context
+        optional_fields = self.request.QUERY_PARAMS.get('fields', [])
+        optional_fields_list = optional_fields.split(',') if optional_fields else []
+        context['fields'] = DEFAULT_FIELDS + [field for field in optional_fields_list if field in OPTIONAL_FIELDS]
+        return context
+
+    def get_queryset(self):
+        course_id = self.request.QUERY_PARAMS.get('course_id', None)
+
+        try:
```

Comment:
The `course_id` is optional. If it is not present all bookmarks for the user should be returned.